// ghost-cursor/lib/spoof.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCursor = exports.GhostCursor = exports.getElementBox = exports.getRandomPagePoint = exports.getCDPClient = exports.installMouseHelper = void 0;
exports.path = path;
var debug_1 = __importDefault(require("debug"));
var math_1 = require("./math");
var mouse_helper_1 = require("./mouse-helper");
Object.defineProperty(exports, "installMouseHelper", { enumerable: true, get: function () { return mouse_helper_1.installMouseHelper; } });
var log = (0, debug_1.default)('ghost-cursor');
/** Helper function to wait a specified number of milliseconds  */
var delay = function (ms) { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                if (ms < 1)
                    return [2 /*return*/];
                return [4 /*yield*/, new Promise(function (resolve) { return setTimeout(resolve, ms); })];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
/**
 * Calculate the amount of time needed to move from (x1, y1) to (x2, y2)
 * given the width of the element being clicked on
 * https://en.wikipedia.org/wiki/Fitts%27s_law
 */
var fitts = function (distance, width) {
    var a = 0;
    var b = 2;
    var id = Math.log2(distance / width + 1);
    return a + b * id;
};
/** Get a random point on a box */
var getRandomBoxPoint = function (_a, options) {
    var x = _a.x, y = _a.y, width = _a.width, height = _a.height;
    var paddingWidth = 0;
    var paddingHeight = 0;
    if ((options === null || options === void 0 ? void 0 : options.paddingPercentage) !== undefined &&
        (options === null || options === void 0 ? void 0 : options.paddingPercentage) > 0 &&
        (options === null || options === void 0 ? void 0 : options.paddingPercentage) <= 100) {
        paddingWidth = (width * options.paddingPercentage) / 100;
        paddingHeight = (height * options.paddingPercentage) / 100;
    }
    return {
        x: x + paddingWidth / 2 + Math.random() * (width - paddingWidth),
        y: y + paddingHeight / 2 + Math.random() * (height - paddingHeight)
    };
};
/** The function signature to access the internal CDP client changed in puppeteer 14.4.1 */
var getCDPClient = function (page) {
    return typeof page._client === 'function'
        ? page._client()
        : page._client;
};
exports.getCDPClient = getCDPClient;
/** Get a random point on a browser window */
var getRandomPagePoint = function (page) { return __awaiter(void 0, void 0, void 0, function () {
    var targetId, window;
    var _a, _b;
    return __generator(this, function (_c) {
        switch (_c.label) {
            case 0:
                targetId = page.target()._targetId;
                return [4 /*yield*/, (0, exports.getCDPClient)(page).send('Browser.getWindowForTarget', { targetId: targetId })];
            case 1:
                window = _c.sent();
                return [2 /*return*/, getRandomBoxPoint({
                        x: math_1.origin.x,
                        y: math_1.origin.y,
                        width: (_a = window.bounds.width) !== null && _a !== void 0 ? _a : 0,
                        height: (_b = window.bounds.height) !== null && _b !== void 0 ? _b : 0
                    })];
        }
    });
}); };
exports.getRandomPagePoint = getRandomPagePoint;
/** Get correct position of Inline elements (elements like `<a>`). Has fallback. */
var getElementBox = function (page_1, element_1) {
    var args_1 = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        args_1[_i - 2] = arguments[_i];
    }
    return __awaiter(void 0, __spreadArray([page_1, element_1], __read(args_1), false), void 0, function (page, element, relativeToMainFrame) {
        var objectId, quads, elementBox, elementFrame, iframes, frame, iframes_1, iframes_1_1, iframe, e_1_1, frameBox, _a, elementBox, _b;
        var e_1, _c;
        var _d;
        if (relativeToMainFrame === void 0) { relativeToMainFrame = true; }
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    _e.trys.push([0, 14, , 20]);
                    objectId = element.remoteObject().objectId;
                    if (objectId === undefined)
                        throw new Error('Element objectId is undefined, falling back to alternative methods');
                    return [4 /*yield*/, (0, exports.getCDPClient)(page).send('DOM.getContentQuads', { objectId: objectId })];
                case 1:
                    quads = _e.sent();
                    elementBox = {
                        x: quads.quads[0][0],
                        y: quads.quads[0][1],
                        width: quads.quads[0][4] - quads.quads[0][0],
                        height: quads.quads[0][5] - quads.quads[0][1]
                    };
                    if (!!relativeToMainFrame) return [3 /*break*/, 13];
                    return [4 /*yield*/, element.contentFrame()];
                case 2:
                    elementFrame = _e.sent();
                    return [4 /*yield*/, ((_d = elementFrame === null || elementFrame === void 0 ? void 0 : elementFrame.parentFrame()) === null || _d === void 0 ? void 0 : _d.$$('xpath/.//iframe'))];
                case 3:
                    iframes = _e.sent();
                    if (!(iframes !== undefined && iframes !== null)) return [3 /*break*/, 13];
                    frame = void 0;
                    _e.label = 4;
                case 4:
                    _e.trys.push([4, 9, 10, 11]);
                    iframes_1 = __values(iframes), iframes_1_1 = iframes_1.next();
                    _e.label = 5;
                case 5:
                    if (!!iframes_1_1.done) return [3 /*break*/, 8];
                    iframe = iframes_1_1.value;
                    return [4 /*yield*/, iframe.contentFrame()];
                case 6:
                    if ((_e.sent()) === elementFrame) {
                        frame = iframe;
                    }
                    _e.label = 7;
                case 7:
                    iframes_1_1 = iframes_1.next();
                    return [3 /*break*/, 5];
                case 8: return [3 /*break*/, 11];
                case 9:
                    e_1_1 = _e.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 11];
                case 10:
                    try {
                        if (iframes_1_1 && !iframes_1_1.done && (_c = iframes_1.return)) _c.call(iframes_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 11:
                    if (!(frame !== undefined && frame != null)) return [3 /*break*/, 13];
                    return [4 /*yield*/, frame.boundingBox()];
                case 12:
                    frameBox = _e.sent();
                    if (frameBox !== null) {
                        elementBox.x -= frameBox.x;
                        elementBox.y -= frameBox.y;
                    }
                    _e.label = 13;
                case 13: return [2 /*return*/, elementBox];
                case 14:
                    _a = _e.sent();
                    _e.label = 15;
                case 15:
                    _e.trys.push([15, 17, , 19]);
                    log('Quads not found, trying regular boundingBox');
                    return [4 /*yield*/, element.boundingBox()];
                case 16:
                    elementBox = _e.sent();
                    if (elementBox === null)
                        throw new Error('Element boundingBox is null, falling back to getBoundingClientRect');
                    return [2 /*return*/, elementBox];
                case 17:
                    _b = _e.sent();
                    log('BoundingBox null, using getBoundingClientRect');
                    return [4 /*yield*/, element.evaluate(function (el) {
                            return el.getBoundingClientRect();
                        })];
                case 18: return [2 /*return*/, _e.sent()];
                case 19: return [3 /*break*/, 20];
                case 20: return [2 /*return*/];
            }
        });
    });
};
exports.getElementBox = getElementBox;
/** Generates a set of points for mouse movement between two coordinates. */
function path(start, end, 
/**
 * Additional options for generating the path.
 * Can also be a number which will set `spreadOverride`.
 */
// TODO: remove number arg in next major version change, fine to just allow `spreadOverride` in object.
options) {
    var optionsResolved = typeof options === 'number'
        ? { spreadOverride: options }
        : __assign({}, options);
    var DEFAULT_WIDTH = 100;
    var MIN_STEPS = 25;
    var width = 'width' in end && end.width !== 0 ? end.width : DEFAULT_WIDTH;
    var curve = (0, math_1.bezierCurve)(start, end, optionsResolved.spreadOverride);
    var length = curve.length() * 0.8;
    var speed = optionsResolved.moveSpeed !== undefined && optionsResolved.moveSpeed > 0
        ? (25 / optionsResolved.moveSpeed)
        : Math.random();
    var baseTime = speed * MIN_STEPS;
    var steps = Math.ceil((Math.log2(fitts(length, width) + 1) + baseTime) * 3);
    var re = curve.getLUT(steps);
    return clampPositive(re, optionsResolved);
}
var clampPositive = function (vectors, options) {
    var clampedVectors = vectors.map(function (vector) { return ({
        x: Math.max(0, vector.x),
        y: Math.max(0, vector.y)
    }); });
    return (options === null || options === void 0 ? void 0 : options.useTimestamps) === true ? generateTimestamps(clampedVectors, options) : clampedVectors;
};
var generateTimestamps = function (vectors, options) {
    var _a;
    var speed = (_a = options === null || options === void 0 ? void 0 : options.moveSpeed) !== null && _a !== void 0 ? _a : (Math.random() * 0.5 + 0.5);
    var timeToMove = function (P0, P1, P2, P3, samples) {
        var total = 0;
        var dt = 1 / samples;
        for (var t = 0; t < 1; t += dt) {
            var v1 = (0, math_1.bezierCurveSpeed)(t * dt, P0, P1, P2, P3);
            var v2 = (0, math_1.bezierCurveSpeed)(t, P0, P1, P2, P3);
            total += (v1 + v2) * dt / 2;
        }
        return Math.round(total / speed);
    };
    var timedVectors = [];
    for (var i = 0; i < vectors.length; i++) {
        if (i === 0) {
            timedVectors.push(__assign(__assign({}, vectors[i]), { timestamp: Date.now() }));
        }
        else {
            var P0 = vectors[i - 1];
            var P1 = vectors[i];
            var P2 = i + 1 < vectors.length ? vectors[i + 1] : (0, math_1.extrapolate)(P0, P1);
            var P3 = i + 2 < vectors.length ? vectors[i + 2] : (0, math_1.extrapolate)(P1, P2);
            var time = timeToMove(P0, P1, P2, P3, vectors.length);
            timedVectors.push(__assign(__assign({}, vectors[i]), { timestamp: timedVectors[i - 1].timestamp + time }));
        }
    }
    return timedVectors;
};
var shouldOvershoot = function (a, b, threshold) {
    return (0, math_1.magnitude)((0, math_1.direction)(a, b)) > threshold;
};
var intersectsElement = function (vec, box) {
    return (vec.x > box.x &&
        vec.x <= box.x + box.width &&
        vec.y > box.y &&
        vec.y <= box.y + box.height);
};
var GhostCursor = /** @class */ (function () {
    function GhostCursor(page, _a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.start, start = _c === void 0 ? math_1.origin : _c, _d = _b.performRandomMoves, performRandomMoves = _d === void 0 ? false : _d, _e = _b.defaultOptions, defaultOptions = _e === void 0 ? {} : _e, _f = _b.visible, visible = _f === void 0 ? false : _f;
        /** Whether mouse is moving via `click`, `move`, `moveTo`, etc. (does not include random movements). Initial state: not moving. */
        this.moving = false;
        this.page = page;
        this.location = start;
        this.defaultOptions = defaultOptions;
        if (visible) {
            // Install mouse helper (visible mouse). Do not await the promise but return immediately
            this.installMouseHelper().then(function (_) { }, function (_) { });
        }
        // Start random mouse movements. Do not await the promise but return immediately
        if (performRandomMoves) {
            this.randomMove().then(function (_) { }, function (_) { });
        }
    }
    /**
     * Install mouse helper (visible cursor).
     */
    GhostCursor.prototype.installMouseHelper = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, mouse_helper_1.installMouseHelper)(this.page).then(function (_a) {
                            var removeMouseHelper = _a.removeMouseHelper;
                            _this.removeMouseHelperFn = removeMouseHelper;
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Make the cursor no longer visible.
     * Only has an effect if `visible=true` was passed, or this.installMouseHelper performed manually.
     */
    GhostCursor.prototype.removeMouseHelper = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, ((_a = this.removeMouseHelperFn) === null || _a === void 0 ? void 0 : _a.call(this))];
                    case 1:
                        _b.sent();
                        this.removeMouseHelperFn = undefined;
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Move the mouse to a point, getting the vectors via `path(previous, newLocation, options)`  */
    GhostCursor.prototype.moveMouse = function (newLocation_1, options_1) {
        return __awaiter(this, arguments, void 0, function (newLocation, options, abortOnMove) {
            var cdpClient, vectors, vectors_1, vectors_1_1, v, dispatchParams, error_1, e_2_1;
            var e_2, _a;
            if (abortOnMove === void 0) { abortOnMove = false; }
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        cdpClient = (0, exports.getCDPClient)(this.page);
                        vectors = path(this.location, newLocation, options);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 8, 9, 10]);
                        vectors_1 = __values(vectors), vectors_1_1 = vectors_1.next();
                        _b.label = 2;
                    case 2:
                        if (!!vectors_1_1.done) return [3 /*break*/, 7];
                        v = vectors_1_1.value;
                        _b.label = 3;
                    case 3:
                        _b.trys.push([3, 5, , 6]);
                        // In case this is called from random mouse movements and the users wants to move the mouse, abort
                        if (abortOnMove && this.moving) {
                            return [2 /*return*/];
                        }
                        dispatchParams = {
                            type: 'mouseMoved',
                            x: v.x,
                            y: v.y
                        };
                        if ('timestamp' in v)
                            dispatchParams.timestamp = v.timestamp;
                        return [4 /*yield*/, cdpClient.send('Input.dispatchMouseEvent', dispatchParams)];
                    case 4:
                        _b.sent();
                        this.location = v;
                        return [3 /*break*/, 6];
                    case 5:
                        error_1 = _b.sent();
                        // Exit function if the browser is no longer connected
                        if (!this.page.browser().isConnected())
                            return [2 /*return*/];
                        log('Warning: could not move mouse, error message:', error_1);
                        return [3 /*break*/, 6];
                    case 6:
                        vectors_1_1 = vectors_1.next();
                        return [3 /*break*/, 2];
                    case 7: return [3 /*break*/, 10];
                    case 8:
                        e_2_1 = _b.sent();
                        e_2 = { error: e_2_1 };
                        return [3 /*break*/, 10];
                    case 9:
                        try {
                            if (vectors_1_1 && !vectors_1_1.done && (_a = vectors_1.return)) _a.call(vectors_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                        return [7 /*endfinally*/];
                    case 10: return [2 /*return*/];
                }
            });
        });
    };
    /** Start random mouse movements. Function recursively calls itself. */
    GhostCursor.prototype.randomMove = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, rand, _1;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ moveDelay: 2000, randomizeMoveDelay: true }, (_a = this.defaultOptions) === null || _a === void 0 ? void 0 : _a.randomMove), options);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 6, , 7]);
                        if (!!this.moving) return [3 /*break*/, 4];
                        return [4 /*yield*/, (0, exports.getRandomPagePoint)(this.page)];
                    case 2:
                        rand = _b.sent();
                        return [4 /*yield*/, this.moveMouse(rand, optionsResolved, true)];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4: return [4 /*yield*/, delay(optionsResolved.moveDelay * (optionsResolved.randomizeMoveDelay ? Math.random() : 1))];
                    case 5:
                        _b.sent();
                        this.randomMove(options).then(function (_) { }, function (_) { }); // fire and forget, recursive function
                        return [3 /*break*/, 7];
                    case 6:
                        _1 = _b.sent();
                        log('Warning: stopping random mouse movements');
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    GhostCursor.prototype.mouseButtonAction = function (action, options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, cdpClient;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ button: 'left', clickCount: 1 }, (_a = this.defaultOptions) === null || _a === void 0 ? void 0 : _a.click), options);
                        cdpClient = (0, exports.getCDPClient)(this.page);
                        return [4 /*yield*/, cdpClient.send('Input.dispatchMouseEvent', {
                                x: this.location.x,
                                y: this.location.y,
                                button: optionsResolved.button,
                                clickCount: optionsResolved.clickCount,
                                type: action
                            })];
                    case 1:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Mouse button down */
    GhostCursor.prototype.mouseDown = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.mouseButtonAction('mousePressed', options)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Mouse button up (release) */
    GhostCursor.prototype.mouseUp = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.mouseButtonAction('mouseReleased', options)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Toggles random mouse movements on or off. */
    GhostCursor.prototype.toggleRandomMove = function (random) {
        this.moving = !random;
    };
    /** Get current location of the cursor. */
    GhostCursor.prototype.getLocation = function () {
        return this.location;
    };
    /**
     * Simulates a mouse click at the specified selector or element.
     * Default is to click at current location, don't move.
     */
    GhostCursor.prototype.click = function (selector, 
    /** @default defaultOptions.click */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, wasRandom, error_2;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ moveDelay: 2000, hesitate: 0, waitForClick: 0, randomizeMoveDelay: true, button: 'left', clickCount: 1 }, (_a = this.defaultOptions) === null || _a === void 0 ? void 0 : _a.click), options);
                        wasRandom = !this.moving;
                        this.toggleRandomMove(false);
                        if (!(selector !== undefined)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.move(selector, __assign(__assign({}, optionsResolved), { 
                                // apply moveDelay after click, but not after actual move
                                moveDelay: 0 }))];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 7, , 8]);
                        return [4 /*yield*/, delay(optionsResolved.hesitate)];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, this.mouseDown()];
                    case 4:
                        _b.sent();
                        return [4 /*yield*/, delay(optionsResolved.waitForClick)];
                    case 5:
                        _b.sent();
                        return [4 /*yield*/, this.mouseUp()];
                    case 6:
                        _b.sent();
                        return [3 /*break*/, 8];
                    case 7:
                        error_2 = _b.sent();
                        log('Warning: could not click mouse, error message:', error_2);
                        return [3 /*break*/, 8];
                    case 8: return [4 /*yield*/, delay(optionsResolved.moveDelay * (optionsResolved.randomizeMoveDelay ? Math.random() : 1))];
                    case 9:
                        _b.sent();
                        this.toggleRandomMove(wasRandom);
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Moves the mouse to the specified selector or element. */
    GhostCursor.prototype.move = function (selector, 
    /** @default defaultOptions.move */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, wasRandom, go;
            var _this = this;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ moveDelay: 0, maxTries: 10, overshootThreshold: 500, randomizeMoveDelay: true }, (_a = this.defaultOptions) === null || _a === void 0 ? void 0 : _a.move), options);
                        wasRandom = !this.moving;
                        this.toggleRandomMove(false);
                        go = function (iteration) { return __awaiter(_this, void 0, void 0, function () {
                            var elem, box, destination, newBoundingBox;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        if (iteration > (optionsResolved.maxTries)) {
                                            throw Error('Could not mouse-over element within enough tries');
                                        }
                                        return [4 /*yield*/, this.getElement(selector, optionsResolved)
                                            // Make sure the object is in view
                                        ];
                                    case 1:
                                        elem = _a.sent();
                                        // Make sure the object is in view
                                        return [4 /*yield*/, this.scrollIntoView(elem, optionsResolved)];
                                    case 2:
                                        // Make sure the object is in view
                                        _a.sent();
                                        return [4 /*yield*/, (0, exports.getElementBox)(this.page, elem)];
                                    case 3:
                                        box = _a.sent();
                                        destination = (optionsResolved.destination !== undefined)
                                            ? (0, math_1.add)(box, optionsResolved.destination)
                                            : getRandomBoxPoint(box, optionsResolved);
                                        if (!shouldOvershoot(this.location, destination, optionsResolved.overshootThreshold)) return [3 /*break*/, 6];
                                        // overshoot
                                        return [4 /*yield*/, this.moveMouse((0, math_1.overshoot)(destination, GhostCursor.OVERSHOOT_RADIUS), optionsResolved)
                                            // then go to the box
                                        ];
                                    case 4:
                                        // overshoot
                                        _a.sent();
                                        // then go to the box
                                        return [4 /*yield*/, this.moveMouse(__assign(__assign({}, box), destination), __assign(__assign({}, optionsResolved), { spreadOverride: GhostCursor.OVERSHOOT_SPREAD }))];
                                    case 5:
                                        // then go to the box
                                        _a.sent();
                                        return [3 /*break*/, 8];
                                    case 6: 
                                    // go directly to the box, no overshoot
                                    return [4 /*yield*/, this.moveMouse(destination, optionsResolved)];
                                    case 7:
                                        // go directly to the box, no overshoot
                                        _a.sent();
                                        _a.label = 8;
                                    case 8: return [4 /*yield*/, (0, exports.getElementBox)(this.page, elem)
                                        // It's possible that the element that is being moved towards
                                        // has moved to a different location by the time
                                        // the the time the mouseover animation finishes
                                    ];
                                    case 9:
                                        newBoundingBox = _a.sent();
                                        if (!!intersectsElement(this.location, newBoundingBox)) return [3 /*break*/, 11];
                                        return [4 /*yield*/, go(iteration + 1)];
                                    case 10: return [2 /*return*/, _a.sent()];
                                    case 11: return [2 /*return*/];
                                }
                            });
                        }); };
                        return [4 /*yield*/, go(0)];
                    case 1:
                        _b.sent();
                        this.toggleRandomMove(wasRandom);
                        return [4 /*yield*/, delay(optionsResolved.moveDelay * (optionsResolved.randomizeMoveDelay ? Math.random() : 1))];
                    case 2:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Moves the mouse to the specified destination point. */
    GhostCursor.prototype.moveTo = function (destination, 
    /** @default defaultOptions.moveTo */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, wasRandom;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ moveDelay: 0, randomizeMoveDelay: true }, (_a = this.defaultOptions) === null || _a === void 0 ? void 0 : _a.moveTo), options);
                        wasRandom = !this.moving;
                        this.toggleRandomMove(false);
                        return [4 /*yield*/, this.moveMouse(destination, optionsResolved)];
                    case 1:
                        _b.sent();
                        this.toggleRandomMove(wasRandom);
                        return [4 /*yield*/, delay(optionsResolved.moveDelay * (optionsResolved.randomizeMoveDelay ? Math.random() : 1))];
                    case 2:
                        _b.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Moves the mouse by a specified amount */
    GhostCursor.prototype.moveBy = function (delta, options) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.moveTo((0, math_1.add)(this.location, __assign({ x: 0, y: 0 }, delta)), options)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Scrolls the element into view. If already in view, no scroll occurs. */
    GhostCursor.prototype.scrollIntoView = function (selector, 
    /** @default defaultOptions.scroll */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, scrollSpeed, elem, _a, viewportWidth, viewportHeight, docHeight, docWidth, scrollPositionTop, scrollPositionLeft, elemBoundingBox, elemBox, marginedBox, marginedBoxRelativeToDoc, targetBox, top, left, bottom, right, isInViewport, manuallyScroll, cdpClient, objectId, _b, e_3;
            var _this = this;
            var _c;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ scrollDelay: 200, scrollSpeed: 100, inViewportMargin: 0 }, (_c = this.defaultOptions) === null || _c === void 0 ? void 0 : _c.scroll), options);
                        scrollSpeed = (0, math_1.clamp)(optionsResolved.scrollSpeed, 1, 100);
                        return [4 /*yield*/, this.getElement(selector, optionsResolved)];
                    case 1:
                        elem = _d.sent();
                        return [4 /*yield*/, this.page.evaluate(function () { return ({
                                viewportWidth: document.body.clientWidth,
                                viewportHeight: document.body.clientHeight,
                                docHeight: document.body.scrollHeight,
                                docWidth: document.body.scrollWidth,
                                scrollPositionTop: window.scrollY,
                                scrollPositionLeft: window.scrollX
                            }); })];
                    case 2:
                        _a = _d.sent(), viewportWidth = _a.viewportWidth, viewportHeight = _a.viewportHeight, docHeight = _a.docHeight, docWidth = _a.docWidth, scrollPositionTop = _a.scrollPositionTop, scrollPositionLeft = _a.scrollPositionLeft;
                        return [4 /*yield*/, (0, exports.getElementBox)(this.page, elem)]; // is relative to viewport
                    case 3:
                        elemBoundingBox = _d.sent() // is relative to viewport
                        ;
                        elemBox = {
                            top: elemBoundingBox.y,
                            left: elemBoundingBox.x,
                            bottom: elemBoundingBox.y + elemBoundingBox.height,
                            right: elemBoundingBox.x + elemBoundingBox.width
                        };
                        marginedBox = {
                            top: elemBox.top - optionsResolved.inViewportMargin,
                            left: elemBox.left - optionsResolved.inViewportMargin,
                            bottom: elemBox.bottom + optionsResolved.inViewportMargin,
                            right: elemBox.right + optionsResolved.inViewportMargin
                        };
                        marginedBoxRelativeToDoc = {
                            top: marginedBox.top + scrollPositionTop,
                            left: marginedBox.left + scrollPositionLeft,
                            bottom: marginedBox.bottom + scrollPositionTop,
                            right: marginedBox.right + scrollPositionLeft
                        };
                        targetBox = {
                            top: Math.max(marginedBoxRelativeToDoc.top, 0) - scrollPositionTop,
                            left: Math.max(marginedBoxRelativeToDoc.left, 0) - scrollPositionLeft,
                            bottom: Math.min(marginedBoxRelativeToDoc.bottom, docHeight) - scrollPositionTop,
                            right: Math.min(marginedBoxRelativeToDoc.right, docWidth) - scrollPositionLeft
                        };
                        top = targetBox.top, left = targetBox.left, bottom = targetBox.bottom, right = targetBox.right;
                        isInViewport = top >= 0 &&
                            left >= 0 &&
                            bottom <= viewportHeight &&
                            right <= viewportWidth;
                        if (isInViewport)
                            return [2 /*return*/];
                        manuallyScroll = function () { return __awaiter(_this, void 0, void 0, function () {
                            var deltaY, deltaX;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        deltaY = 0;
                                        deltaX = 0;
                                        if (top < 0) {
                                            deltaY = top; // Scroll up
                                        }
                                        else if (bottom > viewportHeight) {
                                            deltaY = bottom - viewportHeight; // Scroll down
                                        }
                                        if (left < 0) {
                                            deltaX = left; // Scroll left
                                        }
                                        else if (right > viewportWidth) {
                                            deltaX = right - viewportWidth; // Scroll right
                                        }
                                        return [4 /*yield*/, this.scroll({ x: deltaX, y: deltaY }, optionsResolved)];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); };
                        _d.label = 4;
                    case 4:
                        _d.trys.push([4, 13, , 15]);
                        cdpClient = (0, exports.getCDPClient)(this.page);
                        if (!(scrollSpeed === 100 && optionsResolved.inViewportMargin <= 0)) return [3 /*break*/, 10];
                        _d.label = 5;
                    case 5:
                        _d.trys.push([5, 7, , 9]);
                        objectId = elem.remoteObject().objectId;
                        if (objectId === undefined)
                            throw new Error();
                        return [4 /*yield*/, cdpClient.send('DOM.scrollIntoViewIfNeeded', { objectId: objectId })];
                    case 6:
                        _d.sent();
                        return [3 /*break*/, 9];
                    case 7:
                        _b = _d.sent();
                        return [4 /*yield*/, manuallyScroll()];
                    case 8:
                        _d.sent();
                        return [3 /*break*/, 9];
                    case 9: return [3 /*break*/, 12];
                    case 10: return [4 /*yield*/, manuallyScroll()];
                    case 11:
                        _d.sent();
                        _d.label = 12;
                    case 12: return [3 /*break*/, 15];
                    case 13:
                        e_3 = _d.sent();
                        // use regular JS scroll method as a fallback
                        log('Falling back to JS scroll method', e_3);
                        return [4 /*yield*/, elem.evaluate(function (e) { return e.scrollIntoView({
                                block: 'center',
                                behavior: scrollSpeed < 90 ? 'smooth' : undefined
                            }); })];
                    case 14:
                        _d.sent();
                        return [3 /*break*/, 15];
                    case 15: return [2 /*return*/];
                }
            });
        });
    };
    /** Scrolls the page the distance set by `delta`. */
    GhostCursor.prototype.scroll = function (delta, 
    /** @default defaultOptions.scroll */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, scrollSpeed, cdpClient, deltaX, deltaY, xDirection, yDirection, largerDistanceDir, _a, largerDistance, shorterDistance, EXP_SCALE_START, largerDistanceScrollStep, numSteps, largerDistanceRemainder, shorterDistanceScrollStep, shorterDistanceRemainder, i, longerDistanceDelta, shorterDistanceDelta, _b, deltaX_1, deltaY_1;
            var _c, _d, _e;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ scrollDelay: 200, scrollSpeed: 100 }, (_c = this.defaultOptions) === null || _c === void 0 ? void 0 : _c.scroll), options);
                        scrollSpeed = (0, math_1.clamp)(optionsResolved.scrollSpeed, 1, 100);
                        cdpClient = (0, exports.getCDPClient)(this.page);
                        deltaX = (_d = delta.x) !== null && _d !== void 0 ? _d : 0;
                        deltaY = (_e = delta.y) !== null && _e !== void 0 ? _e : 0;
                        xDirection = deltaX < 0 ? -1 : 1;
                        yDirection = deltaY < 0 ? -1 : 1;
                        deltaX = Math.abs(deltaX);
                        deltaY = Math.abs(deltaY);
                        largerDistanceDir = deltaX > deltaY ? 'x' : 'y';
                        _a = __read(largerDistanceDir === 'x' ? [deltaX, deltaY] : [deltaY, deltaX], 2), largerDistance = _a[0], shorterDistance = _a[1];
                        EXP_SCALE_START = 90;
                        largerDistanceScrollStep = scrollSpeed < EXP_SCALE_START
                            ? scrollSpeed
                            : (0, math_1.scale)(scrollSpeed, [EXP_SCALE_START, 100], [EXP_SCALE_START, largerDistance]);
                        numSteps = Math.floor(largerDistance / largerDistanceScrollStep);
                        largerDistanceRemainder = largerDistance % largerDistanceScrollStep;
                        shorterDistanceScrollStep = Math.floor(shorterDistance / numSteps);
                        shorterDistanceRemainder = shorterDistance % numSteps;
                        i = 0;
                        _f.label = 1;
                    case 1:
                        if (!(i < numSteps)) return [3 /*break*/, 4];
                        longerDistanceDelta = largerDistanceScrollStep;
                        shorterDistanceDelta = shorterDistanceScrollStep;
                        if (i === numSteps - 1) {
                            longerDistanceDelta += largerDistanceRemainder;
                            shorterDistanceDelta += shorterDistanceRemainder;
                        }
                        _b = __read(largerDistanceDir === 'x'
                            ? [longerDistanceDelta, shorterDistanceDelta]
                            : [shorterDistanceDelta, longerDistanceDelta], 2), deltaX_1 = _b[0], deltaY_1 = _b[1];
                        deltaX_1 = deltaX_1 * xDirection;
                        deltaY_1 = deltaY_1 * yDirection;
                        return [4 /*yield*/, cdpClient.send('Input.dispatchMouseEvent', {
                                type: 'mouseWheel',
                                deltaX: deltaX_1,
                                deltaY: deltaY_1,
                                x: this.location.x,
                                y: this.location.y
                            })];
                    case 2:
                        _f.sent();
                        _f.label = 3;
                    case 3:
                        i++;
                        return [3 /*break*/, 1];
                    case 4: return [4 /*yield*/, delay(optionsResolved.scrollDelay)];
                    case 5:
                        _f.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Scrolls to the specified destination point. */
    GhostCursor.prototype.scrollTo = function (destination, 
    /** @default defaultOptions.scroll */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, _a, docHeight, docWidth, scrollPositionTop, scrollPositionLeft, to;
            var _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        optionsResolved = __assign(__assign({ scrollDelay: 200, scrollSpeed: 100 }, (_b = this.defaultOptions) === null || _b === void 0 ? void 0 : _b.scroll), options);
                        return [4 /*yield*/, this.page.evaluate(function () { return ({
                                docHeight: document.body.scrollHeight,
                                docWidth: document.body.scrollWidth,
                                scrollPositionTop: window.scrollY,
                                scrollPositionLeft: window.scrollX
                            }); })];
                    case 1:
                        _a = _c.sent(), docHeight = _a.docHeight, docWidth = _a.docWidth, scrollPositionTop = _a.scrollPositionTop, scrollPositionLeft = _a.scrollPositionLeft;
                        to = (function () {
                            switch (destination) {
                                case 'top':
                                    return { y: 0 };
                                case 'bottom':
                                    return { y: docHeight };
                                case 'left':
                                    return { x: 0 };
                                case 'right':
                                    return { x: docWidth };
                                default:
                                    return destination;
                            }
                        })();
                        return [4 /*yield*/, this.scroll({
                                y: to.y !== undefined ? to.y - scrollPositionTop : 0,
                                x: to.x !== undefined ? to.x - scrollPositionLeft : 0
                            }, optionsResolved)];
                    case 2:
                        _c.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /** Gets the element via a selector. Can use an XPath. */
    GhostCursor.prototype.getElement = function (selector, 
    /** @default defaultOptions.getElement */
    options) {
        return __awaiter(this, void 0, void 0, function () {
            var optionsResolved, elem, _a, handle;
            var _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        optionsResolved = __assign(__assign({}, (_b = this.defaultOptions) === null || _b === void 0 ? void 0 : _b.getElement), options);
                        elem = null;
                        if (!(typeof selector === 'string')) return [3 /*break*/, 9];
                        if (!(selector.startsWith('//') || selector.startsWith('(//'))) return [3 /*break*/, 4];
                        selector = "xpath/.".concat(selector);
                        if (!(optionsResolved.waitForSelector !== undefined)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.page.waitForSelector(selector, { timeout: optionsResolved.waitForSelector })];
                    case 1:
                        _c.sent();
                        _c.label = 2;
                    case 2: return [4 /*yield*/, this.page.$$(selector)];
                    case 3:
                        _a = __read.apply(void 0, [_c.sent(), 1]), handle = _a[0];
                        elem = handle.asElement();
                        return [3 /*break*/, 8];
                    case 4:
                        if (!(optionsResolved.waitForSelector !== undefined)) return [3 /*break*/, 6];
                        return [4 /*yield*/, this.page.waitForSelector(selector, { timeout: optionsResolved.waitForSelector })];
                    case 5:
                        _c.sent();
                        _c.label = 6;
                    case 6: return [4 /*yield*/, this.page.$(selector)];
                    case 7:
                        elem = _c.sent();
                        _c.label = 8;
                    case 8:
                        if (elem === null) {
                            throw new Error("Could not find element with selector \"".concat(selector, "\", make sure you're waiting for the elements by specifying \"waitForSelector\""));
                        }
                        return [3 /*break*/, 10];
                    case 9:
                        // ElementHandle
                        elem = selector;
                        _c.label = 10;
                    case 10: return [2 /*return*/, elem];
                }
            });
        });
    };
    GhostCursor.OVERSHOOT_SPREAD = 10;
    GhostCursor.OVERSHOOT_RADIUS = 120;
    return GhostCursor;
}());
exports.GhostCursor = GhostCursor;
/**
 * @deprecated
 * TODO: Remove on next major version change. Prefer to just do `new GhostCursor` instead of this function.
 * Is here because removing would be breaking.
 */
var createCursor = function (page, 
/**
 * Cursor start position.
 * @default { x: 0, y: 0 }
 */
start, 
/**
 * Initially perform random movements.
 * If `move`,`click`, etc. is performed, these random movements end.
 * @default false
 */
performRandomMoves, 
/**
 * Default options for cursor functions.
 */
defaultOptions, 
/**
 * Whether cursor should be made visible using `installMouseHelper`.
 * @default false
 */
visible) {
    if (start === void 0) { start = math_1.origin; }
    if (performRandomMoves === void 0) { performRandomMoves = false; }
    if (defaultOptions === void 0) { defaultOptions = {}; }
    if (visible === void 0) { visible = false; }
    return new GhostCursor(page, { start: start, performRandomMoves: performRandomMoves, defaultOptions: defaultOptions, visible: visible });
};
exports.createCursor = createCursor;
